tests/integration.rs

```diff
@@ -0,0 +1,433 @@
+                .arg("-dnsserver")
+                .arg("127.0.0.1:8053") // Matched to default -dns01 addr for pebble-challtestsrv.
+                .arg("-strict"),
+        )?;
+
+        let challtestsrv = Subprocess::new(
+            Command::new(&challtestsrv_path)
+                .arg("-doh-cert")
+                .arg("tests/testdata/server.pem")
+                .arg("-doh-cert-key")
+                .arg("tests/testdata/server.key"),
+        )?;
+
+        wait_for_server(&config.pebble.listen_address);
+
+        Ok(Self {
+            config,
+            config_file,
+            pebble,
+            challtestsrv,
+        })
+    }
+
+    // Return an HTTP client configured to trust the Pebble management interface root CA.
+    //
+    // Note: this is distinct from the CA issuer.
+    fn http_client(&self) -> HyperClient<hyper_rustls::HttpsConnector<HttpConnector>, Full<Bytes>> {
+        let mut roots = RootCertStore::empty();
+        roots.add_parsable_certificates(
+            CertificateDer::pem_file_iter("tests/testdata/ca.pem")
+                .unwrap()
+                .map(|result| result.unwrap()),
+        );
+
+        HyperClient::builder(TokioExecutor::new()).build(
+            hyper_rustls::HttpsConnectorBuilder::new()
+                .with_tls_config(
+                    rustls::ClientConfig::builder()
+                        .with_root_certificates(roots)
+                        .with_no_client_auth(),
+                )
+                .https_or_http()
+                .enable_http1()
+                .enable_http2()
+                .build(),
+        )
+    }
+
+    // Provision an HTTP-01 challenge response for the given token and key authorization.
+    //
+    // The Pebble challenge test server will be configured to respond to HTTP-01 challenge
+    // requests for the provided token by returning the expected key auth.
+    async fn add_http01_response(
+        &self,
+        token: &str,
+        key_auth: &str,
+    ) -> Result<(), Box<dyn StdError>> {
+        let client = self.http_client();
+
+        let body = json!({
+            "token": token,
+            "content": key_auth,
+        })
+        .to_string();
+
+        let req = Request::builder()
+            .method(Method::POST)
+            .uri(format!("{}/add-http01", self.challenge_management_url()))
+            .header(CONTENT_TYPE, "application/json")
+            .body(Full::new(body.into()))?;
+
+        client.request(req).await?;
+
+        Ok(())
+    }
+
+    // Return a RootCertStore containing the issuer root for the Pebble CA.
+    //
+    // This is distinct from the management root CA, and is randomly generated each
+    // time that Pebble starts up. This is the issuer that signs the randomly generated
+    // intermediate certificate returned as part of ACME issued certificate chains.
+    async fn issuer_roots(&self) -> Result<RootCertStore, Box<dyn StdError>> {
+        let client = self.http_client();
```

Comment:
Nit: `fn http_client()` should come after this method.

Reply:
Done.